client/tinychain/collection/tensor.py

```diff
@@ -635,6 +647,69 @@ def gradients(self, loss):
         return Gradients()
 
 
+class Reduce(Dual):
+    # TODO: move common functionality from reduce operators into this parent class
+    pass
+
+
+class Norm(Dual):
+    def forward(self):
+        return NDArray.norm(self.subject, self.args)
+    
+    def backward(self, variable=None):
+        return self.subject / self.subject.norm(self.args)
+    
+    def gradients(self, loss):
+        from ..ml.optimizer import Variable
+
+        grads = Gradients()
+
+        if isinstance(self.subject, Variable):
+            grads.update(self.subject.invert(loss * self.subject / self.subject.norm(self.args)))
+        elif operator(self.subject):
+            grads.update(operator(self.subject).gradients(loss * self.subject / self.subject.norm(self.args)))
+        
+        return grads
+        
+
+# TODO: make it work for multiple functions and for .sum(axis=tuple)
+class Sum(Reduce):
+    def forward(self):
+        return NDArray.sum(self.subject, axis=self.args)
+
+    def backward(self, variable=None):
+        from tinychain.collection.tensor import Dense
+
+        return Dense.ones_like(self.subject)
```

Comment:
This will only work in the case that `self.subject` is a `Variable`. I think what you want is `return derivative_of(self.subject, variable).sum(self.args)`.